Feature/github issue count by label chart

frenchbread opened the pull request:
Closes #221 

# Proposed changes
- Add bar chart to display repository issue count by label

brylie commented on integrations/github/tags/server/methods.js:
```diff
@@ -0,0 +1,33 @@
+import { Github } from '/integrations/github/server/github';
```
Consider putting this in `/integrations/github/labels/...` since everything else is using the word *labels*.

brylie commented:
@frenchbread please get peer review from @Nazarah regarding the UX/UI.

brylie commented on integrations/github/labels/server/methods.js:
```diff
@@ -0,0 +1,84 @@
+      page++;
+
+      // Stop syncing if no results returned
+      if (labels.length === 0) {
```
Maybe move this to the top of the function. This will improve readability somewhat and prevent the `_.each(labels, ...)` loop from running on an empty array.

frenchbread commented:
@Nazarah, could you please advice in terms of chart placement in the data source table. 
Currently looks like this: 
![screen shot 2016-04-26 at 17 29 20](https://cloud.githubusercontent.com/assets/2122679/14821964/6f7b534e-0bd5-11e6-9d17-edb9b3787174.png)
I was thinking probably to move both charts to the second row, so that their width would be 1/2 of the full table width, e.g:

Source | Owner | Repo | Contribtors
------------ | ------------- | ------------- | ------------- |
... | ... | ... | ...

Commits Acticity | Labels / Issues count
------------ | -------------
... | ...

brylie commented:
Good idea. Lets see how that looks/works. Also, this level of detail may be moved to a 'integration details' view. E.g. we have discussed having a 'view more detail' or a collapsible section (e.g. via [Bootstrap collapse](http://www.w3schools.com/bootstrap/bootstrap_collapse.asp).

frenchbread commented:
I've experimented a bit and it turns out that if there happen to be quite much issue labels, it looks messy:

![screen shot 2016-04-27 at 22 30 15](https://cloud.githubusercontent.com/assets/2122679/14866853/5d7e041c-0cce-11e6-89d0-908adc099a57.png)
Did not find an option to make the chart with dynamic height. Even though it would be a huge list.
Another example with use of labels list (not chart):
![screen shot 2016-04-27 at 22 42 50](https://cloud.githubusercontent.com/assets/2122679/14866875/7668fb08-0cce-11e6-96e6-e9ad6908c333.png)

brylie commented:
@frenchbread how about if the template takes an argument for the number of labels to display. E.g. we would be saying "show the top 5 labels" or "show the top 10 labels" by using:

```
{{> githubLabelHistogram numberOfLabels=5 }}
```

When including the template without any argument, it could default to showing all. That way, we can use it in other contexts.

frenchbread commented:
@brylie Good point! I'll do that

brylie commented:
@frenchbread what is the status of this PR?

frenchbread commented:
Please review. 

I've added `numberOfLabels` parameter for issueLabelsTemplate. If no value is passed then chart shows all labels for a data source (repo).
I also removed direct repo labels sync methods since they are not needed when showing chart. I commented them out for now If we'll need it later. 
There is also "labels list" approach rather then chart commented out. If current chart is confirmed, I can remove all the commented lines.

brylie commented:
OK, when PR is ready for review, you can remove the WIP label, so that it is easy to tell the PR is ready for review.

brylie commented:
Please consider how to do the following:

- Show the five most frequently used tags for a given repository
  - e.g. 500 bugs, 450 enhancements, 300 help wanted, 250 low hanging fruit
  - the chart currently shows five labels, but the labels may not be frequently used
- Show only whole numbers, or integers. 
  - the chart currently shows decimal places and non-integer number
  - there can't really be 'half of an issue'
- prevent axis text from clipping
  - axis text is currently clipped by the edge of the containing `td` element

frenchbread commented:
@brylie For some reason value format at x axis does not work.
I use `chart.xAxis.tickFormat(d3.format("d"));` Any thoughts?

brylie commented:
I will check it out again.

brylie commented:
You might need to [specify a precision](https://github.com/mbostock/d3/wiki/Formatting):

```js
chart.xAxis.tickFormat(d3.format("0f"));
// or
chart.xAxis.tickFormat(d3.format("0d"));
```

brylie commented:
Also, consider using a slightly larger left margin, so the y-axis labels to not get clipped.

brylie commented:
Ah, found the issue. You need to set the chart formatting before you `.call(chart)`

brylie commented:
I have pushed some changes.

Notice that I switched from `chart.xAxis` to `chart.yAxis`... strange.

brylie commented:
@frenchbread regarding the clipped labels, can we
-  increase the left margin slightly
- truncate the labels at a certain length, so that the end of the word is truncated rather than the beginning getting clipped?

frenchbread commented:
@brylie I think "truncate the labels at a certain length, e.g. 7 characters" is better, because for example in apinf repo there are labels with length of a sentence :)

frenchbread commented:
Added fixes, please review.

frenchbread commented:
Now commit activity chart is not shown for some reason..

frenchbread commented:
Done. Everything works now!

brylie commented:
@frenchbread the chart is still clipping around the edges:

![screenshot from 2016-05-03 12-03-53](https://cloud.githubusercontent.com/assets/17307/14978840/44b9f74e-1127-11e6-9c44-38bc22a3b6f8.png)

Specifically,
- we can remove the legend, since there are only *Issues* displaying
- increase the left margin, so that the y-axis labels to not clip at the beginning

brylie commented:
There seems to be 50px padding on the left and right side of the SVG element:

![screenshot from 2016-05-03 12-07-41](https://cloud.githubusercontent.com/assets/17307/14978916/d073ab7c-1127-11e6-83ed-4517aeda652f.png)

frenchbread commented:
@brylie  Actually padding works fine for me but margin does not.
`padding: 0 50px;`
![screen shot 2016-05-03 at 12 33 26](https://cloud.githubusercontent.com/assets/2122679/14979492/59568fec-112b-11e6-9244-c1ce6f7381c6.png)
`margin-left: 50px;`
![screen shot 2016-05-03 at 12 38 25](https://cloud.githubusercontent.com/assets/2122679/14979589/fd8ee636-112b-11e6-81a5-3ef13a149e5c.png)
